### ISSUE TITLE
no-unused-vars does not report error when comma operator is used

### ISSUE THREAD

**Issue by M393** [eslint/eslint#14325]
With `no-unused-vars` enabled eslint does not mark a variable as unused if assignment is combined with the comma operator.

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

Both produce no warning:
```js
let x = 0;
x++, x = 0;
```
```js
let x = 0;
x++, x = 0;
x = 3;
```

**What did you expect to happen?**
It should mark the last occurance of x as an unused variable.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**
eslint thinks everything is fine.

**Steps to reproduce this issue:**
1. [eslint demo](https://eslint.org/demo#eyJ0ZXh0IjoibGV0IHggPSAwO1xueCsrLCB4ID0gMDtcbnggPSAzOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTIsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiY29uc3RydWN0b3Itc3VwZXIiOjIsImZvci1kaXJlY3Rpb24iOjIsImdldHRlci1yZXR1cm4iOjIsIm5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3IiOjIsIm5vLWNhc2UtZGVjbGFyYXRpb25zIjoyLCJuby1jbGFzcy1hc3NpZ24iOjIsIm5vLWNvbXBhcmUtbmVnLXplcm8iOjIsIm5vLWNvbmQtYXNzaWduIjoyLCJuby1jb25zdC1hc3NpZ24iOjIsIm5vLWNvbnN0YW50LWNvbmRpdGlvbiI6Miwibm8tY29udHJvbC1yZWdleCI6Miwibm8tZGVidWdnZXIiOjIsIm5vLWRlbGV0ZS12YXIiOjIsIm5vLWR1cGUtYXJncyI6Miwibm8tZHVwZS1jbGFzcy1tZW1iZXJzIjoyLCJuby1kdXBlLWVsc2UtaWYiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWltcG9ydC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2V0dGVyLXJldHVybiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19)

**Are you willing to submit a pull request to fix this bug?**
No

**Comment by nzakas** [comment]
This is occurring because of the `x++`, not because of the comma operator. The post fix increment operation counts as both a read and a write, which is why there is no warning.

This looks like a bug but I’d like another set of eyes. @eslint/eslint-tsc

**Comment by M393** [comment]
The increment operator also writes (to itself) after reading.
It has to be related to the comma operator in some way.
These are correctly marked:
```js
let x = 0;
x++;
^
```
```js
let x = 0;
x++; x = 0;
     ^
```

**Comment by nzakas** [comment]
It appears you’re right. Here is an example without the increment operator that makes it clear:

```js
let x = 0;
x=2, x = 0;
x = 3;
```

**Comment by snitin315** [comment]
I will take this issue 👍🏻

**Comment by aladdin-add** [comment]
seems it was not exactly related to comma operator.  imagine this code:

```js
var a;
a; // => rvalue
```

`a` will not be marked unused. ([demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOiAyKi9cbnZhciBhO1xuYTtcblxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==)).
Technologically, it's working as expected. the 2nd `a` was used as rvalue (its value was read). 

and these cases can be caught by some other rules:
```js
var a;
a; // => caught by no-unused-expressions
```

```js
var a;
a, b; // caught by no-sequences
```

**Comment by mdjermanovic** [comment]
> and these cases can be caught by some other rules:
> 
> ```js
> var a;
> a; // => caught by no-unused-expressions
> ```

I think it's fine to not account for this in `no-unused-vars`, because `a;` itself is a possible error and it's reasonable to expect that `no-unused-expressions` is enabled.

> ```js
> var a;
> a, b; // caught by no-sequences
> ```

`no-sequences` is mostly stylistic, but this is also caught by `no-unused-expressions`.
